src/azul/service/async_manifest_service.py

```diff
@@ -63,40 +75,93 @@ class AsyncManifestService(AbstractService):
-        self.step_function_helper.start_execution(self.state_machine_name,
-                                                  execution_id,
-                                                  execution_input=input)
-        return Token(execution_id=execution_id,
-                     request_index=0,
-                     wait_time=self._get_next_wait_time(0))
-
-    def inspect_generation(self, token) -> Union[Token, JSON]:
+    def start_generation(self,
+                         input: JSON,
+                         token: Optional[Token] = None,
+                         execution_id: Optional[str] = None,
+                         retry_index: int = 0) -> Token:
+        if execution_id is None:
+            execution_id = input['object_key'].rpartition('/')[2]
+            execution_id = execution_id.partition('.')[0]
+            assert '.' not in execution_id, execution_id
+        if token is not None:
+            execution_id = token.execution_id
+            execution_id = execution_id.partition('_')[0]
+            assert '_' not in execution_id, execution_id
+        execution_id = f'{execution_id}_{retry_index}'
+        if retry_index > self.step_function_execution_retries:
+            raise StateMachineError('Maximum manifest generation retries reached.',
+                                    execution_id)
+        token = Token(execution_id=execution_id,
+                      request_index=0,
+                      wait_time=self._get_next_wait_time(0))
         try:
-            execution = self.step_function_helper.describe_execution(state_machine_name=self.state_machine_name,
-                                                                     execution_name=token.execution_id)
+            self.step_function_helper.start_execution(self.state_machine_name,
+                                                      execution_id,
+                                                      execution_input=input)
         except ClientError as e:
-            if e.response['Error']['Code'] == 'ExecutionDoesNotExist':
-                raise InvalidTokenError from e
+            if e.response['Error']['Code'] == 'ExecutionAlreadyExists':
+                token = self.retry_generation(input, token, execution_id, retry_index)
             else:
                 raise
-        output = execution.get('output', None)
-        status = execution['status']
-        if status == 'SUCCEEDED':
+        return token
+
+    def retry_generation(self,
+                         input: JSON,
+                         token: Token,
+                         execution_id: str,
+                         retry_index: int):
+        output, status = self._inspect_generation(token)
+        if status is StepFunctionStatus.running:
+            return token.advance(wait_time=self._get_next_wait_time(token.request_index))
+        elif status in (StepFunctionStatus.failed,
+                        StepFunctionStatus.succeeded,
```

Comment:
This is either wrong or needs an explanatory comment. Intuitively, why would we retry when it succeeded?

Reply:
Consider the scenario where a previously generated manifest object might not be accessible via cache. Either because the object expired or it was deleted. This will lead the service to request the generation of another manifest object. If the infrastructure that generates the `object_key` hasn't been updated since the inaccessible manifest object was created, the same `execution_id` will be used as in the first request. As is specified in [StartExecution API Reference](https://docs.aws.amazon.com/step-functions/latest/apireference/API_StartExecution.html) `Note: …If the execution is closed… it will return a 400 ExecutionAlreadyExists error. Names can be reused after 90 days.`, a status check on the step function execution will tell us that it succeeded, but there will be no cached manifest object to retrieve. Which is why executions that were successful need to be retried.